specs/experimental/op-contracts-manager.md

```diff
@@ -153,13 +166,98 @@ This provides the following benefits:
 
 - Contract addresses for a chain can be derived as a function of chain ID without any RPC calls.
 - Chain ID uniqueness is enforced for free, as a deploy using the same chain ID
-will result in attempting to deploy to the same address, which is prohibited by
-the EVM.
+  will result in attempting to deploy to the same address, which is prohibited by
+  the EVM.
   - This property is contingent on the proxy and `AddressManager` code not
-  changing when OP Contracts Manager is upgraded. Both of these are not planned to
-  change.
+    changing when OP Contracts Manager is upgraded. Both of these are not planned to
+    change.
   - The OP Contracts Manager is not responsible for enforcing chain ID uniqueness, so it is acceptable
-  if this property is not preserved in future versions of the OP Contracts Manager.
+    if this property is not preserved in future versions of the OP Contracts Manager.
+
+## Upgrading
+
+### Interface
+
+#### `upgrade`
+
+The `upgrade` method is used by the Upgrade Controller to upgrade the full set of L1 contracts for
+all chains that it controls.
+
+It has the following interface:
+
+```solidity
+function upgrade(ISystemConfig[] _systemConfigs, NewChainConfig[] _newConfigs) public;
+```
+
+For each chain successfully upgraded, the following event is emitted:
+
+```solidity
+event Upgraded(uint256 indexed l2ChainId, SystemConfig indexed systemConfig);
+```
+
+This method reverts if the upgrade is not successful for any of the chains.
+
+### Implementation
+
+The high level logic of the upgrade method is as follows:
+
+1. The Upgrade Controller Safe will `DELEGATECALL` to the `OPCM.upgrade()` method.
+2. For each `_systemConfig`, the list of addresses in the chain is retrieved.
+3. For each address, a two step upgrade is used where:
+   1. the first upgrade is to an `InitializerResetter` which resets the `initialized` value.
+   1. the implementation is updated to the final address and `upgrade()` is called on that address.
+
+This approach requires that all contracts have an `upgrade()` function which sets the `initialized`
+value to `true`. The `upgrade` function body should be empty unless it is used to set a new state
+variable added to that contract since the last upgrade.
+
+#### `NewChainConfig` struct
+
+This struct is used to pass the new chain configuration to the `upgrade` method, and so it will
+vary for each release of the OP Contracts Manager, based on what (if any) new parameters are added.
+
+In practice, this struct is likely to be have a unique name for each release of the OP Contracts
+Manager.
+
+#### Requirements on the OP Chain contracts
+
+In general, all contracts used in an OP Chain SHOULD be proxied with a single shared implementation.
+This means that all values which are not constant across OP Chains SHOULD be held in storage rather
+than the bytecode of the implementation.
+
+Any contracts which do not meet this requirement will need to be deployed by the `upgrade()`
```

Comment:
Do we have an example of a contract that exists today which falls into this category? By this I mean, do some contracts in the OP contract ecosystem have require immutable variables which differ between op chains AND theis data can't be held in storage for whatever reason?

Reply:
The `FaultDisputeGame` and `PermissionedDisputeGame` contracts currently contain chain specific immutables ([ex](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol#L124). 

There is [work](https://github.com/ethereum-optimism/design-docs/pull/161) being done to address that, so this may not be an issue, but it seems worthwhile to document how it would be handled if not.

Reply:
Nice.